File: C99/src/lexer.c
//
// Created by irwin on 18/05/2021.
//
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include "../header/lexer.h"

// isDigit
int isDigit(char chr) {
    return '0' <= chr && chr <= '9';
}
// isAlpha
int isAlpha(char chr) {
    return 'a' <= chr && chr <= 'z' || 'A' <= chr && chr <= 'Z';
}
// isIdent
int isIdent(char chr) {
    return isAlpha(chr) || isDigit(chr) || chr == '_';
}
// isSpace
int isSpace(char chr) {
    return chr == ' ' || chr == '\t' || chr == '\n';
}
// advance
void advance(Lexer* lexer) {
    lexer->pos += 1;
    if (lexer->pos > strlen(lexer->input)) {
        lexer->current_char = EOF;
    } else {
        lexer->current_char = lexer->input[lexer->pos];
    }
}
// skipWhitespace
void skipWhitespace(Lexer *lexer) {
    while (lexer->current_char != EOF && isSpace(lexer->current_char)) {
        advance(lexer);
    }
}
// getNumber
Token getNumber(Lexer *lexer) {
    int start_position = lexer->pos;
    while (lexer->current_char != EOF && isDigit(lexer->current_char)) {
        advance(lexer);
    }
    // check for floating point
    if (lexer->current_char == '.') {
        advance(lexer); // skip the '.'
        while (lexer->current_char != EOF && isDigit(lexer->current_char)) {
            advance(lexer);
        }
    }

    int nLength = lexer->pos - start_position;
    char *lexeme = malloc(nLength + 1);
    memcpy(lexeme, lexer->input + start_position, nLength);
    if (lexeme == NULL) {
        printf("Unavailable memory for use.\n");
        exit(1);
    }
    lexeme[nLength] = 0;

    return newToken(TT_NUMBER, lexeme);
}
// getString
Token getString(Lexer *lexer, char strDelim) {
    advance(lexer); // skip first delim
    int start_position = lexer->pos;
    while (lexer->current_char != EOF && lexer->current_char != strDelim) {
        advance(lexer);
    }

    int nLength = lexer->pos - start_position;
    char *lexeme = malloc(nLength + 1);
    if (lexeme == NULL) {
        printf("Unavailable memory for use.\n");
        exit(1);
    }
    memcpy(lexeme, lexer->input + start_position, nLength);
    lexeme[nLength] = '\0'; // close the array string.
    advance(lexer); // skip last delim
    return newToken(TT_STRING, lexeme);
}
// getIdentifier
Token getIdentifier(Lexer *lexer) {
    int start_position = lexer->pos;
    while (lexer->current_char != EOF && isIdent(lexer->current_char)) {
        advance(lexer);
    }
    int nLength = lexer->pos - start_position;
    char *lexeme = malloc(nLength + 1);
    if (lexeme == NULL) {
        printf("Unavailable memory for use.\n");
        exit(1);
    }
    // copy memory from input
    memcpy(lexeme, lexer->input + start_position, nLength);
    lexeme[nLength] = '\0'; // EOF character
    TokenType type = isKeyword(lexeme);

    return newToken(type, lexeme);
}
// nextToken
Token nextToken(Lexer *lexer) {
    while (lexer->current_char) {
        if (isSpace(lexer->current_char)) {
            skipWhitespace(lexer);
            continue;
        }
        if (isDigit(lexer->current_char))
            return getNumber(lexer);
        if (lexer->current_char == '\"' || lexer->current_char == '\'')
            return getString(lexer, lexer->current_char);
        if (isIdent(lexer->current_char))
            return getIdentifier(lexer);
        // 1 length characters
        if (lexer->current_char == '+') {
            advance(lexer);
            return newToken(TT_PLUS, "+");
        }
        if (lexer->current_char == '-') {
            advance(lexer);
            return newToken(TT_MINUS, "-");
        }
        if (lexer->current_char == '*') {
            advance(lexer);
            return newToken(TT_MUL, "*");
        }
        if (lexer->current_char == '/') {
            advance(lexer);
            return newToken(TT_DIV, "/");
        }
        if (lexer->current_char == ',') {
            advance(lexer);
            return newToken(TT_COMMA, ",");
        }
        if (lexer->current_char == ';') {
            advance(lexer);
            return newToken(TT_SEMICOLON, ";");
        }
        if (lexer->current_char == ':') {
            advance(lexer);
            return newToken(TT_COLON, ":");
        }
        if (lexer->current_char == '[') {
            advance(lexer);
            return newToken(TT_LBRACKET, "[");
        }
        if (lexer->current_char == ']') {
            advance(lexer);
            return newToken(TT_RBRACKET, "]");
        }
        if (lexer->current_char == '{') {
            advance(lexer);
            return newToken(TT_LBRACE, "{");
        }
        if (lexer->current_char == '}') {
            advance(lexer);
            return newToken(TT_RBRACE, "}");
        }
        if (lexer->current_char == '(') {
            advance(lexer);
            return newToken(TT_LPAREN, "(");
        }
        if (lexer->current_char == ')') {
            advance(lexer);
            return newToken(TT_RPAREN, ")");
        }
        if (lexer->current_char == '.') {
            advance(lexer);
            return newToken(TT_DOT, ".");
        }
        // double digit characters
        if (lexer->current_char == '<') {
            advance(lexer);
            if (lexer->current_char == '=') {
                advance(lexer);
                return newToken(TT_LESS_EQ, "<=");
            }
            return newToken(TT_LESS, "<");
        }
        if (lexer->current_char == '>') {
            advance(lexer);
            if (lexer->current_char == '=') {
                advance(lexer);
                return newToken(TT_GREATER_EQ, ">=");
            }
            return newToken(TT_GREATER, ">");
        }
        if (lexer->current_char == '!') {
            advance(lexer);
            if (lexer->current_char == '=') {
                advance(lexer);
                return newToken(TT_NOT_EQ, "!=");
            }
            return newToken(TT_NOT, "!");
        }
        if (lexer->current_char == '=') {
            advance(lexer);
            if (lexer->current_char == '=') {
                advance(lexer);
                return newToken(TT_EQUAL, "==");
            }
            return newToken(TT_ASSIGN, "=");
        }
        printf("lexer.c: unknown character '%c'", lexer->current_char);
        exit(1);
    }
    return newToken(TT_EOF, "");
}
// newLexer
Lexer *newLexer(const char* input) {
    Lexer *lexer = (Lexer*)malloc(sizeof(Lexer)); // casted to lexer
    if (!lexer) {
        printf("lexer.c: No available memory for use\n");
        exit(0);
    }
    // update the lexer pointer
    lexer->pos = 0;
    lexer->input = input;
    lexer->current_char = lexer->input[lexer->pos];

    return lexer;
}
// deleteLexer
void deleteLexer(Lexer *lexer) {
    free(lexer);
}

File: Go/backup/lexer/lexer.go
package lexer

import (
	"CornyLang/token"
	"fmt"
	"os"
)

type Lexer struct {
	input        string
	pos          int
	current_char byte
}

func New(input string) *Lexer {
	lexer := &Lexer{input: input, pos: 0}
	lexer.current_char = lexer.input[lexer.pos]
	return lexer
}

func isLetter(chr byte) bool {
	return isAlpha(chr) || isDigit(chr) || chr == '_'
}

func isAlpha(chr byte) bool {
	return 'a' <= chr && chr <= 'z' || 'A' <= chr && chr <= 'Z'
}

func isSpace(chr byte) bool {
	return chr == ' ' || chr == '\t' || chr == '\n'
}

func isDigit(chr byte) bool {
	return '0' <= chr && chr <= '9'
}

func (l *Lexer) advance() {
	l.pos += 1
	if l.pos >= len(l.input) {
		l.current_char = 0
	} else {
		l.current_char = l.input[l.pos]
	}
}

func (l *Lexer) skipWhiteSpace() {
	for isSpace(l.current_char) {
		l.advance()
	}
}

func (l *Lexer) getNumber() token.Token {
	var start_position = l.pos

	for l.current_char != 0 && isDigit(l.current_char) {
		l.advance()
	}
	if l.current_char == '.' {
		l.advance()
		for l.current_char != 0 && isDigit(l.current_char) {
			l.advance()
		}
	}

	return token.Token{Type: token.TT_NUMBER, Literal: l.input[start_position:l.pos]}
}

func (l *Lexer) getString(strDelim byte) token.Token {
	l.advance() // skip the delimiter
	var start_position = l.pos
	for l.current_char != 0 && l.current_char != strDelim {
		l.advance()
	}
	var lexeme = l.input[start_position:l.pos]
	l.advance() // skip ending delimiter

	return token.Token{Type: token.TT_STRING, Literal: lexeme}
}

func (l *Lexer) getIdentifier() token.Token {
	var start_position = l.pos
	for l.current_char != 0 && isLetter(l.current_char) {
		l.advance()
	}
	var lexeme = l.input[start_position:l.pos]
	var tokenType = token.IsKeyword(lexeme)
	return token.Token{Type: tokenType, Literal: lexeme}
}

func (l *Lexer) NextToken() token.Token {
	for l.current_char != 0 {
		if isSpace(l.current_char) {
			l.skipWhiteSpace()
			continue
		}
		if isDigit(l.current_char) {
			return l.getNumber()
		}
		if isLetter(l.current_char) {
			return l.getIdentifier()
		}
		if l.current_char == '"' || l.current_char == '\'' {
			return l.getString(l.current_char)
		}
		if l.current_char == '+' {
			l.advance()
			return newToken(token.TT_PLUS, "+")
		}
		if l.current_char == '-' {
			l.advance()
			return newToken(token.TT_MINUS, "-")
		}
		if l.current_char == '*' {
			l.advance()
			return newToken(token.TT_MUL, "*")
		}
		if l.current_char == '/' {
			l.advance()
			return newToken(token.TT_DIV, "/")
		}
		if l.current_char == ',' {
			l.advance()
			return newToken(token.TT_COMMA, ",")
		}
		if l.current_char == ';' {
			l.advance()
			return newToken(token.TT_SEMICOLON, ";")
		}
		if l.current_char == ':' {
			l.advance()
			return newToken(token.TT_COLON, ":")
		}
		if l.current_char == '[' {
			l.advance()
			return newToken(token.TT_LBRACKET, "[")
		}
		if l.current_char == ']' {
			l.advance()
			return newToken(token.TT_RBRACKET, "]")
		}
		if l.current_char == '{' {
			l.advance()
			return newToken(token.TT_LBRACE, "{")
		}
		if l.current_char == '}' {
			l.advance()
			return newToken(token.TT_RBRACE, "}")
		}
		if l.current_char == '(' {
			l.advance()
			return newToken(token.TT_LPAREN, "(")
		}
		if l.current_char == ')' {
			l.advance()
			return newToken(token.TT_RPAREN, ")")
		}
		if l.current_char == '.' {
			l.advance()
			return newToken(token.TT_DOT, ".")
		}
		if l.current_char == '?' {
			l.advance()
			return newToken(token.TT_QUESTION, "?")
		}
		// double digit characters
		if l.current_char == '<' {
			l.advance()
			if l.current_char == '=' {
				l.advance()
				return newToken(token.TT_LESS_EQ, "<=")
			}
			return newToken(token.TT_LESS, "<")
		}
		if l.current_char == '>' {
			l.advance()
			if l.current_char == '=' {
				l.advance()
				return newToken(token.TT_GREATER_EQ, ">=")
			}
			return newToken(token.TT_GREATER, ">")
		}
		if l.current_char == '!' {
			l.advance()
			if l.current_char == '=' {
				l.advance()
				return newToken(token.TT_NOT_EQ, "!=")
			}
			return newToken(token.TT_NOT, "!")
		}
		if l.current_char == '=' {
			l.advance()
			if l.current_char == '=' {
				l.advance()
				return newToken(token.TT_EQUAL, "==")
			}
			return newToken(token.TT_ASSIGN, "=")
		}
		fmt.Printf("unknown character %c", l.current_char)
		os.Exit(4)
	}
	return token.Token{Type: token.TT_EOF, Literal: ""}
}

func newToken(ttype token.TokenType, lexeme string) token.Token {
	return token.Token{Type: ttype, Literal: lexeme}
}
